File: 072_Edit_Distance/072_1.java
class Solution {
    public int minDistance(String word1, String word2) {
        int m = word1.length();
        int n = word2.length();
        int[][] dp = new int[m + 1][n + 1];

        for (int i = 0; i <= m; ++i) {
            for (int j = 0; j <= n; ++j) {
                if (i == 0 && j == 0) dp[i][j] = 0;
                else if (i == 0)      dp[i][j] = j;
                else if (j == 0)      dp[i][j] = i;
                else if (word1.charAt(i - 1) == word2.charAt(j - 1))
                    dp[i][j] = dp[i - 1][j - 1];
                else
                    dp[i][j] = 1 + MIN(dp[i - 1][j - 1], 
                                       dp[i][j - 1], 
                                       dp[i - 1][j]);
            }
        }
        return dp[m][n];
    }

    private int MIN(int a, int b, int c) {
        return Math.min(a, Math.min(b, c));
    }
}

File: 072_Edit_Distance/072_1.cpp
#define min(a, b, c) ((a) < ((b) < (c) ? (b) : (c)) ? (a) : ((b) < (c) ? (b) : (c)));
typedef pair<int, int> P;

class Solution {
public:
    int minDistance(string word1, string word2) {
        return dpFunc(word1, word2, word1.size() - 1, word2.size() - 1);
    }

    int dpFunc(string word1, string word2, int i, int j) {
        if (memo.find(P(i, j)) != memo.end()) return memo[P(i, j)];
        if (i == -1) return j + 1;
        if (j == -1) return i + 1;
        if (word1[i] == word2[j])
            memo[P(i, j)] = dpFunc(word1, word2, i - 1, j - 1);
        else
            memo[P(i, j)] =  1 + min(dpFunc(word1, word2, i - 1, j),
                                     dpFunc(word1, word2, i, j - 1),
                                     dpFunc(word1, word2, i - 1, j - 1));
        return memo[P(i, j)];
    }
private:
    map<P, int> memo;
};